### PR TITLE
Improve Error Messaging for Actions by Using ExecutionContext's FileTable as Single Source of Truth and by Passing FileID to All Children Tokens.

### PR DESCRIPTION
**Pull request by ethanchewy**
Branched off from: https://github.com/actions/runner/pull/557

**This PR enables file names to be included in action error messages.**

Our goal is to ensure that the fileIDs are associated with all the children tokens in the action.yml so that error messages reference the correct action file to allow for easier debugging. Also, we want to make sure that the FileTable is updated correctly throughout the process to ensure that Error messages will reference always the correctly associated file.

In this PR, we also refactor the code in `ActionManifestManager` to allow for passing `fileID`s to the children tokens. Originally, there was a `_fileTable` attribute for `ActionManifestManager` instances as well as a `FileTable` for ExecutionContext instances. To reduce confusion and potential bugs in the future when we have nested actions, we refactored how the `FileTable` is used so that only `ExecutionContext.FileTable` is referenced. In the `ActionManifestManager` to ensure that only 1 fileTable (aka the ExecutionContext's `fileTable`) is referenced. Moreover, we add functionality to ensure that the `FileTable` is updated every time we load an Action yaml. 

**tl;dr ExecutionContext's FileTable is now the "source of truth" for any filetable related matters.**

### Demos:
=> Before error message: https://github.com/ethanchewy/testing-actions/actions/runs/144061679
`(Line: 13, Col: 11): Unexpected value 'run:::::'`
=> After PR error message: https://github.com/ethanchewy/testing-actions/actions/runs/145235133
`ethanchewy/testing-composite-errors/v3/action.yml (Line: 13, Col: 11): Unexpected value 'run:::::'`

**More demos:**
`action.yml`
```
runs:
    using: "composite"
    steps: 
        - run: echo ${{ fromJson("not json") }}
```
**Error Messaging after delayed evaluation:**  
https://github.com/ethanchewy/testing-actions/actions/runs/145233958
`ethanchewy/testing-composite-errors/v4/action.yml (Line: 13, Col: 16): Unexpected symbol: '"not'. Located at position 10 within expression: fromJson("not json"))`

### CODE REVIEW

**Comment by ethanchewy** [comment]
TODO: `git pull master` after merging the Composite Run Steps PR

**Comment by ethanchewy** [comment]
TODO: Test with error messages such as {{ fromJson('not json') }}

**Comment by ethanchewy** [comment]
Nice to have: clean up file path in error messaging

File: src/Runner.Worker/ActionManifestManager.cs [hunk 1]
```diff
@@ -56,22 +54,25 @@ public override void Initialize(IHostContext hostContext)
 
         public ActionDefinitionData Load(IExecutionContext executionContext, string manifestFile)
         {
-            var context = CreateContext(executionContext);
+            var templateContext = CreateContext(executionContext);
```
Comment (ethanchewy) [inline src/Runner.Worker/ActionManifestManager.cs:57]: I updated it from `context` to `templateContext` b/c I thought `context` has a misleading connotation and could lead to mixup with `executionContext`.

File: src/Runner.Worker/ActionManifestManager.cs [hunk 2]
```diff
@@ -470,14 +476,16 @@ public ActionDefinitionData Load(IExecutionContext executionContext, string mani
...
                         throw new ArgumentNullException($"No steps provided.");
                     }
                     else
                     {
                         return new CompositeActionExecutionData()
                         {
                             Steps = stepsLoaded,
-                            Environment = envComposite
+                            Environment = envComposite,
+                            FileID = fileID
```
Comment (ethanchewy) [inline src/Runner.Worker/ActionManifestManager.cs:488]: Remove

**Comment by ethanchewy** [comment]
Ready for review.

File: src/Runner.Worker/ActionManifestManager.cs [hunk 3]
```diff
@@ -56,22 +57,31 @@ public override void Initialize(IHostContext hostContext)
...
         {
-            var context = CreateContext(executionContext);
+            var templateContext = CreateContext(executionContext);
             ActionDefinitionData actionDefinition = new ActionDefinitionData();
+
+            // Clean up file name real quick
+            // Instead of using Regex which can be computationally expensive, 
+            // we can just remove the # of characters from the fileName according to the length of the basePath
+            string basePath = _hostContext.GetDirectory(WellKnownDirectory.Actions);
+            var fileName = manifestFile.Remove(0, basePath.Length + 1);
```
Comment (ethanchewy) [inline src/Runner.Worker/ActionManifestManager.cs:67]: TODO: Add sanity check here to make sure we don't accidentally error out of the filename is smaller than the total path for some reason.

File: src/Runner.Worker/ActionManifestManager.cs [hunk 4]
```diff
@@ -348,8 +358,12 @@ public ActionDefinitionData Load(IExecutionContext executionContext, string mani
             IExecutionContext executionContext,
             TemplateContext context,
             TemplateToken inputsToken,
-            MappingToken envComposite)
+            Int32 fileID,
+            MappingToken envComposite = null
+            )
         {
+            Trace.Info($"COMPOSITE ACTIONS FILEID: {fileID}");
```
Comment (ethanchewy) [inline src/Runner.Worker/ActionManifestManager.cs:365]: TODO: remove

File: src/Sdk/DTPipelines/Pipelines/ObjectTemplating/PipelineTemplateEvaluator.cs [hunk 5]
```diff
@@ -160,16 +160,23 @@ public class PipelineTemplateEvaluator
...
+
+                // Pass original filetable to the new context:
+                foreach (var f in fileTable) {
+                    context.GetFileId(f);
+                }
+
                 try
                 {
-                    token = TemplateEvaluator.Evaluate(context, PipelineTemplateConstants.StepsInTemplate, token, 0, null, omitHeader: true);
+                    token = TemplateEvaluator.Evaluate(context, PipelineTemplateConstants.StepsInTemplate, token, 0, null, omitHeader: false);
```
Comment (ethanchewy) [inline src/Sdk/DTPipelines/Pipelines/ObjectTemplating/PipelineTemplateEvaluator.cs:179]: Double check

**Comment by ethanchewy** [comment]
Ready for review.

File: src/Runner.Worker/ActionManifestManager.cs [hunk 6]
```diff
@@ -33,12 +33,13 @@ public interface IActionManifestManager : IRunnerService
     public sealed class ActionManifestManager : RunnerService, IActionManifestManager
     {
         private TemplateSchema _actionManifestSchema;
-        private IReadOnlyList<String> _fileTable;
-
+        private IHostContext _hostContext;
```
Comment (TingluoHuang) [inline src/Runner.Worker/ActionManifestManager.cs:36]: you don't need this one.

HostContext is in RunnerService.

File: src/Runner.Worker/ActionManifestManager.cs [hunk 7]
```diff
@@ -55,22 +56,35 @@ public override void Initialize(IHostContext hostContext)
...
         public ActionDefinitionData Load(IExecutionContext executionContext, string manifestFile)
         {
-            var context = CreateContext(executionContext);
+            var templateContext = CreateContext(executionContext);
             ActionDefinitionData actionDefinition = new ActionDefinitionData();
+
+            // Clean up file name real quick
+            // Instead of using Regex which can be computationally expensive, 
+            // we can just remove the # of characters from the fileName according to the length of the basePath
+            string basePath = _hostContext.GetDirectory(WellKnownDirectory.Actions);
```
Comment (TingluoHuang) [inline src/Runner.Worker/ActionManifestManager.cs:65]: `HostContext.GetDirectory()`

File: src/Runner.Worker/ActionManifestManager.cs [hunk 8]
```diff
@@ -104,18 +118,18 @@ public ActionDefinitionData Load(IExecutionContext executionContext, string mani
...
             {
-                foreach (var error in context.Errors)
+                foreach (var error in templateContext.Errors)
                 {
                     Trace.Error($"Action.yml load error: {error.Message}");
                     executionContext.Error(error.Message);
                 }
 
-                throw new ArgumentException($"Fail to load {manifestFile}");
+                throw new ArgumentException($"Fail to load {fileName}");
```
Comment (TingluoHuang) [inline src/Runner.Worker/ActionManifestManager.cs:132]: will this always shows `actions.yml`?

File: src/Sdk/DTObjectTemplating/ObjectTemplating/TemplateEvaluator.cs [hunk 9]
```diff
@@ -31,7 +31,6 @@ internal partial class TemplateEvaluator
             Boolean omitHeader = false)
         {
             TemplateToken result;
-
```
Comment (TingluoHuang) [inline src/Sdk/DTObjectTemplating/ObjectTemplating/TemplateEvaluator.cs:34]: undo this. 😆

File: src/Sdk/DTPipelines/Pipelines/ObjectTemplating/PipelineTemplateEvaluator.cs [hunk 10]
```diff
@@ -160,13 +160,20 @@ public class PipelineTemplateEvaluator
         }
 
         public List<ActionStep> LoadCompositeSteps(
-            TemplateToken token)
```
Comment (TingluoHuang) [inline src/Sdk/DTPipelines/Pipelines/ObjectTemplating/PipelineTemplateEvaluator.cs:163]: need to copy this change back to server.

File: src/Runner.Worker/ActionManifestManager.cs [hunk 11]
```diff
@@ -55,22 +56,35 @@ public override void Initialize(IHostContext hostContext)
...
             {
                 var token = default(TemplateToken);
 
                 // Get the file ID
-                var fileId = context.GetFileId(manifestFile);
-                _fileTable = context.GetFileTable();
+                var fileId = templateContext.GetFileId(fileName);
+
+                // Add this file to the FileTable in executionContext
+                executionContext.FileTable.Add(fileName);
```
Comment (ericsciple) [inline src/Runner.Worker/ActionManifestManager.cs:80]: After you add the "fileName" the `fileId` should match `executionContext.FileTable.Length`

Can we add an assert.

Comment (ericsciple) [reply src/Runner.Worker/ActionManifestManager.cs:80]: only add if new

File: src/Runner.Worker/ActionManifestManager.cs [hunk 12]
```diff
@@ -297,7 +311,8 @@ public ActionDefinitionData Load(IExecutionContext executionContext, string mani
         private ActionExecutionData ConvertRuns(
             IExecutionContext executionContext,
             TemplateContext context,
-            TemplateToken inputsToken)
+            TemplateToken inputsToken,
+            Int32 fileID)
```
Comment (ericsciple) [inline src/Runner.Worker/ActionManifestManager.cs:315]: `int`

Comment (ericsciple) [reply src/Runner.Worker/ActionManifestManager.cs:315]: is this used?

File: src/Runner.Worker/Handlers/ScriptHandler.cs [hunk 13]
```diff
@@ -291,7 +291,7 @@ public async Task RunAsync(ActionRunStage stage)
                 // Error
                 if (exitCode != 0)
                 {
-                    ExecutionContext.Error($"Process completed with exit code {exitCode}.");
+                    ExecutionContext.Error($"Process completed with exit code {exitCode}");
```
Comment (ericsciple) [inline src/Runner.Worker/Handlers/ScriptHandler.cs]: +1

File: src/Sdk/DTPipelines/Pipelines/ObjectTemplating/PipelineTemplateEvaluator.cs [hunk 14]
```diff
@@ -160,13 +160,20 @@ public class PipelineTemplateEvaluator
...
+            TemplateToken token,
```
Comment (ericsciple) [inline src/Sdk/DTPipelines/Pipelines/ObjectTemplating/PipelineTemplateEvaluator.cs:163]: dont forget to revert this

File: src/Runner.Worker/ActionManifestManager.cs (hunk 8 above)
Comment (ethanchewy) [reply src/Runner.Worker/ActionManifestManager.cs:132]: TODO for ethan: Change variable name top a more clear name

File: src/Runner.Worker/ActionManifestManager.cs (hunk 6 above)
Comment (ethanchewy) [reply src/Runner.Worker/ActionManifestManager.cs:36]: TODO: we don't need to store _hostContext, we can just use HostContext

File: src/Runner.Worker/ActionManifestManager.cs (hunk 8 above)
Comment (ericsciple) [reply src/Runner.Worker/ActionManifestManager.cs:132]: discussed offline, it's the relative path. Would it help to name the variable `relative___`?

File: src/Runner.Worker/ActionManifestManager.cs [hunk 15]
```diff
@@ -55,22 +53,39 @@ public override void Initialize(IHostContext hostContext)
...
                 var token = default(TemplateToken);
 
                 // Get the file ID
-                var fileId = context.GetFileId(manifestFile);
-                _fileTable = context.GetFileTable();
+                var fileId = templateContext.GetFileId(fileRelativePath);
+
+                // Add this file to the FileTable in executionContext if it hasn't been added already
+                // we use > since fileID is 1 indexed
+                if (fileId > executionContext.FileTable?.Count)
```
Comment (ericsciple) [inline src/Runner.Worker/ActionManifestManager.cs:78]: do we need the `FileTable?` anymore, or should this just be `FileTable` now?

File: src/Runner.Worker/ActionManifestManager.cs [hunk 16]
```diff
@@ -282,12 +297,12 @@ public ActionDefinitionData Load(IExecutionContext executionContext, string mani
                 result.ExpressionFunctions.Add(item);
             }
 
-            // Add the file table
-            if (_fileTable?.Count > 0)
+            // Add the file table from the Execution Context
+            if (executionContext.FileTable?.Count > 0)
```
Comment (ericsciple) [inline src/Runner.Worker/ActionManifestManager.cs:301]: same question here, i think we can drop the `?` now

File: src/Runner.Worker/ActionManifestManager.cs (hunk 15 above)
Comment (ethanchewy) [reply src/Runner.Worker/ActionManifestManager.cs:78]: Yeah, we don't need it anymore since the FileTable should always be non null. Will update this to executionContext.FileTable.Count.

Comment (ethanchewy) [reply src/Runner.Worker/ActionManifestManager.cs:78]: good catch -ty
